src/frontend/src/flows/addDevice/welcomeView/showVerificationCode.ts

```diff
@@ -65,79 +85,114 @@ export const showVerificationCode = async (
   tentativeRegistrationInfo: TentativeRegistrationInfo,
   credentialToBeVerified: CredentialId
 ): Promise<void> => {
-  const container = document.getElementById("pageContent") as HTMLElement;
-  render(pageContent(userNumber, alias, tentativeRegistrationInfo), container);
-  return init(
+  const countdown = AsyncCountdown.fromNanos(
+    tentativeRegistrationInfo.device_registration_timeout
+  );
+
+  showVerificationCodePage({
+    userNumber,
+    alias,
+    tentativeRegistrationInfo,
+    remaining: countdown.remainingFormattedAsync(),
+    cancel: () => {
+      countdown.stop();
+      window.location.reload();
+    },
+  });
+
+  const pollResult = await poll({
     userNumber,
     connection,
-    tentativeRegistrationInfo.device_registration_timeout,
-    credentialToBeVerified
-  );
+    credentialToBeVerified,
+    shouldStop: () => countdown.hasStopped(),
+  });
+
+  return await handlePollResult(userNumber, pollResult);
 };
 
-function poll(
-  userNumber: bigint,
-  connection: Connection,
-  credentialToBeVerified: Array<number>,
-  shouldStop: () => boolean
-): Promise<boolean> {
-  return connection.lookupAuthenticators(userNumber).then((deviceData) => {
-    if (shouldStop()) {
-      return false;
+// Poll until the user number (anchor) contains at least one device with
+// credential id equal to 'credentialToBeVerified', or until `shouldStop`
+// returns `true` for the first time.
+async function poll({
+  userNumber,
+  connection,
+  credentialToBeVerified,
+  shouldStop,
+}: {
+  userNumber: bigint;
+  connection: Connection;
+  credentialToBeVerified: Array<number>;
+  shouldStop: () => boolean;
+}): Promise<"timeout" | "match"> {
+  const verifyCredentials = async (): Promise<boolean> => {
+    let res = undefined;
+    try {
+      res = await anchorHasCredentials({
+        userNumber,
+        credential: credentialToBeVerified,
+        connection,
+      });
+    } catch (e) {
+      // Silently discard the error (though log in console) until we have
+      // support for e.g. toasts
+      console.warn("Error occurred when polling:", e);
     }
-    for (const device of deviceData) {
-      if (device.credential_id.length === 1) {
-        const credentialId = device.credential_id[0];
-
-        if (credentialIdEqual(credentialId, credentialToBeVerified)) {
-          return true;
-        }
-      }
+
+    return res ?? false;
+  };
+
+  while (!shouldStop()) {
```

Comment:
side note: "while something infinite" tends to make me "anxious" they might never stops, I rather like readin recursive function 🤷‍♂️

that said, should we add a max timeout to this loop? like shoult it stops in error after e.g. 1min?

Reply:
speaking of maybe no while or recursive function shall be use, would not this be a use case for using `setinterval`?

Reply:
the loop may need to keep running for 15mn. The previous version was recursive, which just hid the problem. An issue with`setInterval` is that we'd then need to make sure we don't lose the handle in case we abort... all in all, this is the safest I could think off. I'm also a bit hesitant to add a max timeout (like one hour) because anyway 15mn is quite long!

Reply:
thanks for the explanation, no perfect solution at a time. maybe something that websocket on the IC will ultimately solve (don't know the exact use case here). all good anyway then